crates/holochain/src/core/workflow/countersigning_workflow.rs

```diff
@@ -1,332 +1,616 @@
-                    // If this is a countersigning session we can grab all the ops
-                    // for this cells action so we can check if we need to self publish them.
-                    let r: Result<_, _> = txn
-                        .prepare(
-                            "SELECT basis_hash, hash FROM DhtOp WHERE action_hash = :action_hash",
-                        )?
-                        .query_map(
-                            named_params! {
-                                ":action_hash": this_cells_action_hash
-                            },
-                            |row| {
-                                let hash: DhtOpHash = row.get("hash")?;
-                                let basis: OpBasis = row.get("basis_hash")?;
-                                Ok((hash, basis))
-                            },
-                        )?
-                        .collect();
-                    return Ok(r?);
+async fn try_recover_failed_sessions(
+    space: &Space,
+    network: &Arc<impl HolochainP2pDnaT + Sized>,
+    cell_id: CellId,
+    signal_tx: &Sender<Signal>,
+    self_trigger: &TriggerSender,
+) {
+    let sessions_in_unknown_state = space
+        .countersigning_workspace
+        .inner
+        .share_ref(|inner| {
+            Ok(inner
+                .sessions
+                .iter()
+                .filter_map(|(author, session_state)| match session_state {
+                    CountersigningSessionState::Unknown { .. } => Some(author.clone()),
+                    _ => None,
+                })
+                .collect::<Vec<_>>())
+        })
+        .unwrap();
+
+    let mut remaining_sessions_in_unknown_state = 0;
+    for author in sessions_in_unknown_state {
+        tracing::info!(
+            "Countersigning session for agent {:?} is in an unknown state, attempting to resolve",
+            author
+        );
+        match incomplete::inner_countersigning_session_incomplete(
+            space.clone(),
+            network.clone(),
+            author.clone(),
+        )
+        .await
+        {
+            Ok((SessionCompletionDecision::Complete(_), _)) => {
+                // No need to do anything here. Signatures were found which may be able to complete
+                // the session but the session isn't actually complete yet. We need to let the
+                // workflow re-run and try those signatures.
+            }
+            Ok((SessionCompletionDecision::Abandoned, _)) => {
+                // The session state has been resolved, so we can remove it from the workspace.
+                let removed_session = space
+                    .countersigning_workspace
+                    .inner
+                    .share_mut(|inner, _| {
+                        tracing::trace!(
+                            "Decision made for incomplete session, removing from workspace: {:?}",
+                            author
+                        );
```

Comment:
The other traces in here are at `info!` level. If it's getting abandoned, should this also be info?

Reply:
Could do, this is meant to be for somebody debugging the workflow to make sure that the state actually gets removed when it's supposed to be. It's not particularly useful information for the user.